# Ponder and Prove The Pigeonhole Principle
## Due: Saturday, 27 March 2021

## Exploration

You are invited to explore some of the ramifications of the *Pigeonhole Principle*, a fascinating counting principle belonging to combinatorics.


### Requirements

Read Brother Kent Bessey's essay entitled *Pigeons and Pigeonholes*, think about and explore the topics therein, and write your response thereto.

Include in your response a description of a discussion about this essay that you will have had with someone *not in the class* (nor in any of the STEM majors). This person can be a spouse, friend, roommate, or relative. The point is to find someone who is not mathematically strong, and has probably never heard of the Pigeonhole Principle, or combinatorics (or discrete mathematics) in general.

#### Copyright Note

Brother Kent A. Bessey, Professor of Mathematics at BYU-Idaho, holds the copyright to this essay, which means:

> All rights reserved (by him). Reproduced by permission (from him).
Read-only. Please do not copy, print, transmit or save a copy of
this work.

He has authorized your temporary use of his work, which [you can find here](https://firstthreeodds.org/17657741833134731255/pigeons-and-pigeonholes.pdf).

### TODO Concretize the Abstract by Writing Some Code

Specifically, write some Python code to help you figure out the problem on page 40, which Brother Bessey prefaces by saying:

> I leave it to the reader to verify the calculations for a final (albeit challenging) exercise ...!

Include the results of your grappling with this problem, and note the two stubbed functions below that you must flesh out in order to verify the calculations and explain the hows and whys.

#### Stub Code

In [1]:
def calculate_number_of_possible_initials_including_middle_initial_if_any(n):
  # YOUR CODE GOES HERE
  return n * n * (n + 1)

def pigeonhole_from_pigeon(pigeon_represented_as_a_person_with_traits):
  # YOUR CODE GOES HERE
  # Get the initials
  names = (pigeon_represented_as_a_person_with_traits['Name']).split()
  initials = names[0][0].upper()
  if len(names) == 3:
    initials += names[1][0].upper() + names[2][0].upper()
  else:
    initials += ' ' + names[1][0].upper()

  # Get the binaries
  binaries = ''
  binary_keys = list(filter(lambda key: key != 'Name', list(pigeon.keys())))
  for key in binary_keys:
    if pigeon_represented_as_a_person_with_traits[key]:
      binaries += '1'
    else:
      binaries += '0'

  # Assert the string is exactly 10 characters long
  string = initials + ':' + binaries
  assert(len(string) == 10)

  # Assert the cardinality of the combinations of the initials and binaries is over 1 million
  cardinality = calculate_number_of_possible_initials_including_middle_initial_if_any(26) * 2 ** len(binary_keys)
  print(cardinality)
  assert(cardinality > 1000000)

  # Returns the string 'J D:111001'
  return string


In [2]:
pigeon = {'Name': 'John Doe', 'LDS': True, 'Likes Cates': True, 'Born Before': True, 'Wears Contacts': False, 'Bites Fingernails': False, 'Lives Close': True}
pigeonhole = pigeonhole_from_pigeon(pigeon)
print(pigeonhole)

1168128
J D:111001


In [3]:
num_pigeonholes = 26 * 27 * 26 * 2 * 2 * 2 * 2 * 2 * 2
# Population of Idaho: https://www.macrotrends.net/states/idaho/population
num_pigeons = 1826913

print(num_pigeons, num_pigeonholes, num_pigeons / num_pigeonholes)

1826913 1168128 1.5639664488823142


### TODO Find Applications of the Pigeonhole Principle

* In any graph with at least two nodes, there are
at least two nodes of the same degree.

* A graph G consists of a set V (the elements of V are called vertices) and a set E of pairs of vertices
(the elements of E are called edges). The degree of a vertex v in a graph is the number of edges
containing v.


* Showing languages are non-regular with the Pigonhole principle.

  Example : L1 = { 0n1n | n ≥ 0 } is non-regular
  * Assume L1 is regular.
  * Then there is a DFA M = (Q, Σ, δ, q0, F) recognizing L1.
  * Define:
    * Pigeons = all strings in 0*.
    * Holes = states in Q.
  * Put pigeon 0i into hole δ*(q0, 0i
  ), that is, the hole
  corresponding to the state reached by input 0i
  .
  * There are |Q| holes, but > |Q| pigeons (actually, infinitely
  many).
  * So by Pigeonhole Principle, 2 pigeons must be put in
  the same hole, say 0i and 0j with i < j.
  * That is, 0i and 0j lead to the same state.
  * Then since M accepts 0i
  1i
  , it also accepts 0j
  1i
  , which is
  incorrect, contradiction.

## TODO My Report on What I Did and What I Learned

### Fun


I did have fun on this assignmet. I liked that a lot of this material was easier (that the last couple of weeks). I also found the concept of the Pigeonhole Principle interesting which made the paper fun to read.

### New

I had never heard of the Pigeonhole Principle. All of the concepts in this assignment were new to me. I learned about this principle and also how to apply this principle in code.

### Meaningful


Considering that all the concepts were new to me, I think that this Prove was more meaningful to me than most. I feel like I learned a lot of new material and was able to understand the concepts being taught. 

### Other

**Connections:**

- There are many connections made to graph theory with the Pigeonhole principle. I mentioned these in the TODO above.

- This can also be connected to decimals in math. YOu can use the pigeonhole principle to prove that the decimal expansion of a rational number m/n is eventually repeating. 

**Collaborators and Contributions:**
- Clair Hocker
- Daniel Strickland
- Bretton Steiner

We met on zoom and worked on the coding section of this assignment. Bretton shared his screen and typed out the code as we all contributed our ideas. 

**Thoughts about the essay in general:**
- The pigeonhole principle is one of the simplest mathematical ideas formulated.
- The pigonhole principle does not deal with probability or likelihood - upon first impression, thought that it did.
- I was suprised that I had never heard of this principle before due to it's simplicity and application in everyday life. 

**Thoughts about the certainty-not-chance feature of the pigeonhole principle:**

- I agree with the cerainty-not-change feature. 
- The example that helped me understand this best read, ": Among 13 people at least two have the same birthday month.
What are the pigeons and what are the pigeonholes? The pigeons are the
13 people and the pigeonholes the 12 months of the year. Distinct months
cannot be assigned to all 13 individuals. Hence, at least two share the
same birthday month."
- It is certainty-not-change, this is because there are only a finite amount of options avaliable and a finite amount of object or people that fit into these categories.

**Thoughts about the philosophical aspect of the pigeonhole principle per page 41:**
- This philosphical aspect on the bottom of page 41, I found to be confusing. 
- I think about problems I have dealt with that have many solutions, for example, in this class everybody programs differently, and thus provides different coding solutions.
- However, regaurdless of how the code is written, the code output would look the same (if done correctly). This helps me understand the point.
- It's not about how you got the solution, it's about the actual solution alone.

**My discussion with a non-STEM non-classmate:**
- I had a discussion with my Dad who works in the world of buisness. 
- My Dad referenced something called the Pigeon Hole Principle in buisness. This occurs when somebody buckets you into a specific category and you are put in a specific slot based off of people perspective. 
- Reverse this: What would happen if there is one pigeon and four holes?
- If you don't have enough to go around, then you have to assume that certain things share. 
- This can be thought about in terms of supply and demand. If the supply is different than the demand you may be forced to share.

**How I grappled with the problem at the bottom of page 40:**

It wasn't too hard to solve this problem on the bottom of page 40, especially after reviewing the material in class. I used the examples given in the article as a guide. Each of the initals has 26 possibilites (27 in the case of using a space for no middle name) and each of the true or false features have 2 possibilities. By multiplying all of the possibilites together, we get 

$26 * 27 * 26 * 2 * 2 * 2 * 2 * 2 * 2 = 1168128$

It was also easy to find the population of Idaho with the URL I used. Since there are more people of Idaho than pigeonholes then there must be at least two people of Idaho who have exactly the same features.


**Connection I found between the pigeonhole principle and graphs:** 

This is detailed in the TODO section above. I talk about the connection between nodes, vertices, and edges. I also talk about the pumping Lemma and how the pigeonhole principle is used to prove that an expression is not regular.



## TODO What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [4]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = True #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = True #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = True #@param {type:'boolean'}
#@markdown ## What is True about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = True #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = True #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = True #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = True #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = True #@param {type:'boolean'}
#@markdown ### I reported what I thought about the essay in general.
cb08 = True #@param {type:'boolean'}
#@markdown ### I reported what I thought about the certainty-not-chance feature of the pigeonhole principle.
cb09 = True #@param {type:'boolean'}
#@markdown ### I reported what I thought about the philosophical aspect of the pigeonhole principle per page 41.
cb10 = True #@param {type:'boolean'}
#@markdown ### I reported on how my discussion with a non-STEM non-classmate went.
cb11 = True #@param {type:'boolean'}
#@markdown ### I reported on how I grappled with the problem at the bottom of page 40.
cb12 = True #@param {type:'boolean'}
#@markdown ### I reported on the connection I found between the pigeonhole principle and graphs.
cb13 = True #@param {type:'boolean'}
#@markdown ## What is True about my code?
#@markdown ### I implemented the first function (the one with the absurdly long name) in one or two lines of code.
cb14 = True #@param {type:'boolean'}
#@markdown ### I used either the sum rule and product rule together or else just the product rule in the first function.
cb15 = True #@param {type:'boolean'}
#@markdown ### The first function calculates and returns the correct value (see Exercise 250).
cb16 = True #@param {type:'boolean'}
#@markdown ### The second function inputs a dictionary like {'Name': 'John Doe', 'LDS': True, 'Likes Cats': True, 'Born Before': True, 'Wears Contacts': False, 'Bites Fingernails': False, 'Lives Close': True} and returns the string 'J D:111001'.
cb17 = True #@param {type:'boolean'}
#@markdown ### The second function calls the first function with the appropriate argument and uses the result.
cb18 = True #@param {type:'boolean'}
#@markdown ### The second function includes an assert that the length of the string being returned is 10.
cb19 = True #@param {type:'boolean'}
#@markdown ### The second function includes an assert that the cardinality of its codomain (set of possible return values) is over a million.
cb20 = True #@param {type:'boolean'}

## DO NOT CHANGE ANYTHING IN THE NEXT CODE CELL!!
### Delete this cell and the following ones before submitting your work.

In [5]:
points_for_what_I_did = [3]*3
points_for_my_report = [5]*8 + [7] + [6]*2
points_for_my_code = [4]*3 + [5]*4
points = points_for_what_I_did + points_for_my_report + points_for_my_code
# cb is short for checkbox
total = sum(map(lambda n, p: p if eval(f'cb{n:02}') else 0,
                range(len(points)), points))             
total

100

# For graders